### PR TITLE
Enable CDEF and restoration filters for 4:2:2

### LINKED ISSUE #1149: Chroma 422 mode causes enc/dec mismatch and its test cannot detect this
It passes "cargo test --release --features=decode_test chroma_sampling_422_aom",
but encode and decoder mismatches if tried for larger size input like:
./target/release/rav1e https://media.xiph.org/video/derf/y4m/flower_garden_422_ntsc.y4m -o test.ivf -r test_rec.y4m --quantizer 100 --speed=5 --limit=1

FYI, that current chroma 422 and 444 are all encoded as INTRA frames, i.e. inter modes are NOT used.

### PR DESCRIPTION
**Pull request by barrbrain**
As far as I can tell, there are no outstanding issues with this combination.
However, I will attempt to revive the fuzzer to answer this with confidence.
Closes #1149.

### CODE REVIEW

**Comment by barrbrain** [comment]
I breathed life into the rav1e fuzzer, and encountered an `attempt to subtract with overflow` at [src/lrf.rs:1416](https://github.com/xiph/rav1e/blob/b54b8d07/src/lrf.rs#L1416) with `w = 26` and `h = 271`.

**Comment by barrbrain** [comment]
The cause may be at [src/lrf.rs:1368](https://github.com/xiph/rav1e/blob/b54b8d07/src/lrf.rs#L1368):
```rust
uv_unit_size =
  uv_unit_size.min(tile_aligned_uv_h_unit_size.min(tile_aligned_uv_v_unit_size));
```

**Comment by lu-zero** [comment]
It sounds possibly related to https://github.com/xiph/rav1e/issues/2212

**Comment by xiphmont** [comment]
It is, but I'm still figuring out in detail how much overlap there is.

**Comment by rzumer** [comment]
Also see #970.

**Comment by barrbrain** [comment]
The newest commit resolves the desync in #1149. The crash in #2212 occurs independent of enabling CDEF. Further analysis is needed for the decoding error when enabling LRF for 422.

File: src/cdef.rs [hunk 1]
```diff
@@ -535,7 +535,11 @@ pub fn cdef_filter_superblock<T: Pixel, U: Pixel>(
               local_sec_strength = cdef_sec_uv_strength << coeff_shift;
               local_damping -= 1;
               if cdef_pri_uv_strength != 0 {
-                dir as usize
+                if xdec != ydec {
+                  [7, 0, 2, 4, 5, 6, 6, 6][dir as usize]
```
Comment (lu-zero) [inline src/cdef.rs:539]: please add a comment about it.

**Comment by barrbrain** [comment]
Added back enabling of LRF for 4:2:2 now that there is no desync locally or in AWCY.
Added as a separate commit so that we may document the impact.

[AWCY results on derf-720p-422 at the default speed](https://beta.arewecompressedyet.com/?job=enable-restoration-filters-422%402020-05-11T09%3A57%3A13.385Z&job=wip-enable-restoration-filters-422%402020-05-11T14%3A31%3A19.634Z) show a 28% decrease in encoding speed:
|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -1.4957 | -1.2369 | -0.5441 |  -1.0947 | -0.9361 | -0.7724 |    -2.1054 |

**Comment by barrbrain** [comment]
The total effect of enabling both filters for 4:2:2 is substantial:
|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -4.0057 | -0.8299 | -1.6084 |  -3.0774 | -1.1085 | -1.8073 |    -5.0115 |

**Comment by barrbrain** [comment]
Until #2212 is resolved, I will minimise changes to LRF, CDEF or tiling parts of the code to avoid confusing anyone looking at that bug. I agree that they could be more readable with some work.

**Pull request merged**